### ISSUE TITLE
UNNEST should not be allowed within CASE statements

### ISSUE THREAD

**Issue by ThomasRolfsnes-EP** [duckdb/duckdb#13466]
### What happens?

When building a case statement it seems the THEN part is allways evaluated. This became evident when I had `unnest(generate_series(..))` in one of my case-stements, resulting in unexpected output.

### To Reproduce

```sql
select
case
 when True then 1
 when False then unnest(generate_series(0,2))
end as x;
```
Results in:
![CleanShot 2024-08-19 at 10 54 43](https://github.com/user-attachments/assets/b3aa0fee-334a-4477-88a1-b536275548ed)

While I expected:
![CleanShot 2024-08-19 at 10 55 06](https://github.com/user-attachments/assets/6f460016-4b07-4fe5-9665-7b466856a335)

### OS:

aarch64

### DuckDB Version:

1.0.0

### DuckDB Client:

Python

### Full Name:

Thomas Rolfsnes

### Affiliation:

Story House Egmont

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have not tested with any build

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by pkoppstein** [comment]
1. I've reproduced the problem using v1.0.1-dev4226
2. I believe the bug is closely associated with the use of `unnest` here.  For example, there's no problem with the similar query:
```
select case
   when True then [1]
   when False then generate_series(0,2)
end as x;
``

**Comment by ThomasRolfsnes-EP** [comment]
@pkoppstein yes indeed, unnest() surfaces the issue here, since it actually modifies the number of rows in the query it becomes evident that the THEN part was evaluated even though WHEN was false

Another way to prove that this is how duckdb currently works is to just throw in costly calculatations that should not be evaluated, like so:

``` python
# without generate_series
%timeit -n 10000 (duckdb.query("select case when True then [1] end as x"))
# 44.8 µs ± 1.54 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

false_stmt = ["when False then generate_series(0,100000)"]
# 1 generate_series
%timeit -n 10000 (duckdb.query(f"select case when True then [1] {' '.join(false_stmt*1)} end as x"))
# 69.4 µs ± 2.06 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# 10 generate_series
%timeit -n 10000 (duckdb.query(f"select case when True then [1] {' '.join(false_stmt*10)} end as x"))
# 369 µs ± 38.6 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

```

**Comment by ThomasRolfsnes-EP** [comment]
FYI `if()` behaves exactly the same:

```
duckdb.query("""
select
  if(True,1,unnest(generate_series(0,2))) as x
""")

┌───────┐
│   x   │
│ int64 │
├───────┤
│     1 │
│     1 │
│     1 │
└───────┘
```

**Comment by Mytherin** [comment]
Thanks for the report!

`UNNEST` is executed completely separately from the regular `SELECT`, similar to an aggregate, so I would say this is more or less expected. Postgres throws an exception here:

```sql
postgres=# select
case
 when True then 1
 when False then unnest(array[0,1,2])
end as x;
ERROR:  set-returning functions are not allowed in CASE
LINE 4:  when False then unnest(array[0,1,2])
                         ^
-- HINT:  You might be able to move the set-returning function into a LATERAL FROM item.
```

I think that is also sane and we might want to switch to throwing an exception.

> Another way to prove that this is how duckdb currently works is to just throw in costly calculatations that should not be evaluated, like so:

DuckDB short-circuits case statements. In your example this most likely does not happen because the individual case statements are constants (i.e. `generate_series(0, 100000)` is a constant) - and the constant folding optimizer is executed before the case statement short-circuiting. Here's a showcase of the case statement short-circuiting:

```sql
D create table ranges as select 999999999999999 range_bound;
D select case when true then [] else generate_series(0, range_bound) end as r from ranges;
┌─────────┐
│    r    │
│ int64[] │
├─────────┤
│ []      │
└─────────┘

-- or with multiple values:
D create or replace table ranges as select 0 as range_bound union all select 999999999999999 range_bound;
D select case when range_bound > 10 then [] else generate_series(0, range_bound) end as r from ranges;
┌─────────┐
│    r    │
│ int64[] │
├─────────┤
│ [0]     │
│ []      │
└─────────┘

```

**Comment by ThomasRolfsnes-EP** [comment]
@Mytherin 

I see, beeing able to use unnest() as part of the select statement is _nice_ (bigquery also does not allow this). 
So optimally case/if would also short-circuit when unnest is used, but if not viable to implement, an exception should absolutely be raised.  E.g., this query should either return 1, _or_ raise an exception:

```
select
  if(True,1,unnest([1,2,3]))
```

From a users perspective I don't think its intuitive or expected at all that the False part here has any impact on the output.

Thanks!

**Comment by ThomasRolfsnes-EP** [comment]
wow, ok I see, that one is also definitely weird, puts a dent in SQL beeing declarative for me :)

I'm fine with raising an exception in the case of unnest. But I'll leave the use-case I had here in case it might influence discussions down the line. 

For certain values of a given column, I needed to duplicate those rows with an incremental counter in a new column, e.g:

```
with input as (
  select 1 as x
  union all
  select 5 as x
)
select
  x,
  case
    when x = 1 then x
    when x > 1 then unnest(generate_series(0,x))
  end as y
from input 
```
![CleanShot 2024-08-23 at 10 11 12](https://github.com/user-attachments/assets/c4ad7033-501b-4228-9155-d0376aa63b0d)

**Comment by Mytherin** [comment]
Perhaps that can be solved by moving the `case` statement into the `unnest`, e.g.:

```sql
with input as (
  select 1 as x
  union all
  select 5 as x
)
select
  x,
  unnest(case
    when x = 1 then [x]
    when x > 1 then generate_series(0,x)
  end) as y
from input ;
```

**Comment by ThomasRolfsnes-EP** [comment]
It does indeed! Nice one, thanks :)

**Comment by pkoppstein** [comment]
@Mytherin - You changed the title from

CASE-statement does not short-circuit UNNEST evaluation

to 

UNNEST should not be allowed within CASE statements

But is that the appropriate conclusion?  Consider:
```
select case when True then 1 else (select * from generate_series(3, 6)) end ;

├───────────────────────────────────────────────────────────────────────────────────────────┤
│                                                                                         1 │
└───────────────────────────────────────────────────────────────────────────────────────────┘

D select case when 1=2 then 1 else (select * from generate_series(3, 6)) end ;

├──────────────────────────────────────────────────────────────────────────────┤
│                                                                            3 │
└──────────────────────────────────────────────────────────────────────────────┘
D 
```

This seems quite correct: 

(1) the logic of `CASE` governs the results;
(2) if one of the branches of the `CASE` statement yields a sequence of values, select the first.

**Comment by soerenwolfers** [comment]
Related: https://github.com/duckdb/duckdb/issues/14012 (which should be titled something like "Errors during constant folding should be caught and only be raised where required" )

**Comment by soerenwolfers** [comment]
Another user with the same problem regarding the `CASE` behavior: https://github.com/duckdb/duckdb/discussions/14188

**Comment by soerenwolfers** [comment]
@Mytherin If you think there is a chance DuckDB could be changed to disallow this like PG, could you reopen this? Very similar confusion befell myself and others over on discord https://discord.com/channels/909674491309850675/1431288775534973060
